src/openforms/submissions/form_logic.py

```diff
@@ -134,72 +114,97 @@ def evaluate_form_logic(
-                variable_key = action["variable"]
-                updated_step_data[variable_key] = new_value
-                data_container.update_data(updated_step_data)
+                data_container.update({action["variable"]: new_value})
+            else:
+                operation = compile_action_operation(action)
+                mutation_operations.append(operation)
+
+    # 6. The variable state is now completely resolved - we can start processing the
+    # dynamic configuration and side effects.
+
+    # Calculate which data has changed from the initial, for the step.
+    updated_step_data = data_container.get_updated_step_data(step)
     step.data = DirtyData(updated_step_data)
 
+    # 7. finally, apply the dynamic configuration
+
+    # we need to apply the context-specific configurations before we can apply
+    # mutations based on logic, which is then in turn passed to the serializer(s)
+    # TODO: refactor this to rely on variables state, which will move this down to
+    configuration = deepcopy(configuration)
+    configuration = get_dynamic_configuration(
+        configuration,
+        # context is expected to contain request, as is the default behaviour with DRF
+        # view(set)s and serializers. Note that :func:`get_dynamic_configuration` is
+        # planned for refactor as part of #1068, which should drop the ``request``
+        # argument. The required information is available on the ``submission`` object
+        # already.
+        request=context.get("request"),
+        submission=submission,
+    )
+
+    # 7.1 Apply the component mutation operations
+    for mutation in mutation_operations:
+        mutation.apply(step, configuration)
+
+    # 7.2 Interpolate the component configuration with the variables.
+    # TODO!
+
+    # 7.3 Handle custom formio types - TODO: this needs to be lifted out of
+    # :func:`get_dynamic_configuration` so that it can use variables.
+
     # Logging the rules
+    initial_data = data_container.initial_data
     logevent.submission_logic_evaluated(
         submission,
         evaluated_rules,
-        {**data_container.static_data, **data_container.initial_data},
+        initial_data,
+        data_container.data,
     )
 
+    # process the output for logic checks with dirty data
     if dirty:
         # only keep the changes in the data, so that old values do not overwrite otherwise
         # debounced client-side data changes
         data_diff = {}
+
         for key, new_value in step.data.items():
-            original_value = data_container.initial_data.get(key)
+            original_value = initial_data.get(key)
             # Reset the value of any field that may have become hidden again after evaluating the logic
             if original_value:
                 component = get_component(configuration, key)
+                default = get_component_default_value(component)
```

Comment:
Does this handle this case? https://github.com/open-formulieren/open-forms/issues/1755

Reply:
I don't know - it might, it might not, but there's a separate PR for that in progress already so we'll see how they merge and if it's still required or not.